### PR TITLE
Feature: Support redirecting to an external URL on successful booking

### LINKED ISSUE #1457: Redirect to URL after successful booking
Let users choose a URL to go to after a successful booking. Maybe also pass some data about the booking in query params (e.g. ?date=1234)

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

Fixes #1457 #2088

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How should this be tested?

- Ensure only URL starting with http:// https:// are accepted even if a direct hit to mutation goes without browser validation.
- Upgrade Modal Box
    - Shown on Input click and Pro Badge Click. Pro Badge shown in Free account only. Trial and Pro doesn’t have it.
- Redirection Test
    - Automatic Redirect after 10 seconds
    - Cross Button to cancel automatic redirection. User stays on the success page and toast is dismissed
    - Continue button takes to external URL immediately.
    - Redirect URL can have query params or hash fragment and they are forwarded as is.
    - Mobile Device
        - Redirecting to {URL} is visible but timer isn’t there considering limited space.
- Team Event -  Redirect can be added and redirection works [Trial And PRO].
    - From **Trial** to **Free** downgrade Team Event keeps working(I would have expected it to stop working) and thus redirection keeps working with it.
    - From **Pro** to **Free** - Same thing
- User Event - Redirect can be added and redirection works.[Trial And PRO]
    - From **Trial** to **Free** downgrade - Redirect stops working but event booking is still working. User can’t edit the redirect URL but he would be able to see the already added URL which won’t work. Pro Upgrade Modal would be shown.
    - From **Pro** to **Free** downgrade - Same thing.

## Checklist

- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/prisma/migrations/20220308100145_/migration.sql [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ALTER TABLE "EventType" ADD COLUMN     "successRedirect" TEXT;
+
+-- RenameIndex
+ALTER INDEX "Booking.uid_unique" RENAME TO "Booking_uid_key";
+
+-- RenameIndex
+ALTER INDEX "DailyEventReference_bookingId_unique" RENAME TO "DailyEventReference_bookingId_key";
+
+-- RenameIndex
+ALTER INDEX "DestinationCalendar.bookingId_unique" RENAME TO "DestinationCalendar_bookingId_key";
```
Comment (baileypumfleet) [inline packages/prisma/migrations/20220308100145_/migration.sql:11]: For some reason, Prisma keeps generating these ALTER INDEX lines, it happened on my last PR

**Review by alannnc — APPROVED** [review]
It worked great!  It only leaves successRedirectUrl with a lot of url query params. Should we clear all of this info from the final url?
image
@zomars @baileypumfleet 

Also can we improve a hint or a label here that says https:// in the url its required? I would change "Redirect on successful booking" to "Redirect on successful booking(https)"
image

File: packages/prisma/migrations/20220308100145_/migration.sql [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ALTER INDEX "VerificationRequest.token_unique" RENAME TO "VerificationRequest_token_key";
+
+-- RenameIndex
+ALTER INDEX "Webhook.id_unique" RENAME TO "Webhook_id_key";
+
+-- RenameIndex
+ALTER INDEX "users.email_unique" RENAME TO "users_email_key";
+
+-- RenameIndex
+ALTER INDEX "users.username_unique" RENAME TO "users_username_key";
```
Comment (zomars) [inline packages/prisma/migrations/20220308100145_/migration.sql:44]:
```suggestion
```

**Review by pumfleet — APPROVED** [review]

**Comment by hariombalhara** [comment]
> It worked great! It only leaves successRedirectUrl with a lot of url query params. Should we clear all of this info from the final url? image @zomars @baileypumfleet
> 
I believe all this info should be kept as is so that whatever info we need to show that success page, owner can also use that to build a very customized success message using that. It also means that we should very well document those query parms


> Also can we improve a hint or a label here that says https:// in the url its required? I would change "Redirect on successful booking" to "Redirect on successful booking(https)" image

Updated placeholder for the hint

**Comment by agustif** [comment]
Wouldn't this new feature allow any bad actor redirect a booking user to any badUrl after booking? Should we at least add some frontend notice that you're getting out of [cal.com](http://cal.com/)?

**Review by zomars — DISMISSED** [review]
I'm blocking this PR for a little bit more so we can address these Qs:

- Can this be used maliciously?
- Wouldn't this reopen a previously fixed vulnerability when you can use Cal to redirect to a malicious site?
- If that's the case how could we handle that?
- And lastly how does the redirect going to work for embedded calendars? (current frames)

**Comment by PeerRich** [comment]
> I'm blocking this PR for a little bit more so we can address these Qs:
> 
> * Can this be used maliciously?
> * Wouldn't this reopen a previously fixed vulnerability when you can use Cal to redirect to a malicious site?
> * If that's the case how could we handle that?
> * And lastly how does the redirect going to work for embedded calendars? (current frames)

I agree with your assessment. We also want to show the success page at least for a couple seconds. I recommend doing this:

image

```
<div class="fixed top-4 inset-x-0 pb-2 sm:pb-5 z-30" style="">
  <div class="max-w-7xl mx-auto px-2 sm:px-6 lg:px-8">
    <div class="p-2 rounded-sm bg-red-600 shadow-lg sm:p-3 bg-green-500">
      <div class="flex items-center justify-between flex-wrap">
        <div class="w-0 flex-1 flex items-center">
          <p class="ml-3 font-medium text-white truncate">
            <span class="md:hidden">Redirecting ...</span>
            <span class="hidden md:inline">You are being redirected to https://external.website.com in 9 seconds.</span>
          </p>
        </div>
        <div class="order-3 mt-2 flex-shrink-0 w-full sm:order-2 sm:mt-0 sm:w-auto">
          <a href="#" class="flex items-center justify-center px-4 py-2 border border-transparent rounded-sm shadow-sm text-sm font-medium text-indigo-600 bg-white hover:bg-indigo-50">Continue</a>
        </div>
        <div class="order-2 flex-shrink-0 sm:order-3 sm:ml-2">
          <button type="button" class="-mr-1 flex p-2 rounded-md hover:bg-indigo-500 focus:outline-none focus:ring-2 focus:ring-white">
            <span class="sr-only">Dismiss</span>
            <!-- Heroicon name: outline/x -->
            <svg class="h-6 w-6 text-white" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" aria-hidden="true">
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12"></path>
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
```

**Comment by PeerRich** [comment]
this was also planned as a PRO feature, which means it needs to be in `/ee/` and disabled for FREE (show the same badge as we do for custom theme)

**Review by PeerRich — CHANGES_REQUESTED** [review]
https://github.com/calcom/cal.com/pull/2087#issuecomment-1066094263

**Comment by PeerRich** [comment]
another problem: when embedding in an iframe, it'll show the new redirect in an iframe.

we need to open in a new tab (until we have the script embed)

**Comment by hariombalhara** [comment]
> another problem: when embedding in an iframe, it'll show the new redirect in an iframe.
> 
> we need to open in a new tab (until we have the script embed)

Attempt to opening a new tab on timer would be blocked by browser. For now I have opened in the same tab but in the top frame.

**Comment by hariombalhara** [comment]
@PeerRich @zomars  Some testing pending otherwise it is ready for review with the suggested changes.

**Comment by alannnc** [comment]
@hariombalhara  
It says I'm missing this lib function 
`import { isSuccessRedirectAvailable } from "@lib/isSuccessRedirectAvailable";`
Couldn’t find any isSuccessRedirectAvailable declaration

**Review by alannnc — DISMISSED** [review]
isSuccessRedirectAvailable function declaration

**Comment by hariombalhara** [comment]
> isSuccessRedirectAvailable function declaration

Oops! Forgot to commit it

**Comment by hariombalhara** [comment]
@alannnc Sorry about that missing file error. 
Testing is complete now from my side. Added manual tests that I have done.

**Comment by lharries** [comment]
Excited to see this PR! 

We are currently working on a feature where we want to redirect the page based on a Cal.com booking being completed within an iFrame.

Does this PR enable an iFrame to trigger the redirection of the top level page? If not, I'd love to request it as a feature (ideally as part of this PR if not out of scope).

As a side point - the embedded Calendly iFrame uses `window.postMessage()` which gives the developer lots of flexibility including using dynamic success URLs and triggering appropriate analytics. Here's a link to how [Calendly does it](https://help.calendly.com/hc/en-us/articles/223147027-Embed-options-overview?tab=advanced#6)

<details>
<summary>Image of relevant Calendly docs</summary>
Screenshot Embed options overview – Help Center (Google Chrome) 2022-03-15 at 10 35
</details>

**Comment by PeerRich** [comment]
> Excited to see this PR!
> 
> We are currently working on a feature where we want to redirect the page based on a Cal.com booking being completed within an iFrame.
> 
> Does this PR enable an iFrame to trigger the redirection of the top level page? If not, I'd love to request it as a feature (ideally as part of this PR if not out of scope).
> 
> For implementation, one way could be to use `window.postMessage()` and allow the developer the handle the rest. This is what the embedded Calendly iFrame uses and gives the developer lots of flexibility including using dynamic success URLs and triggering appropriate analytics. Here's a link to how [Calendly does it](https://help.calendly.com/hc/en-us/articles/223147027-Embed-options-overview?tab=advanced#6)
> 
> Image of relevant Calendly docs

I havent tested but afaik it won't redirect at the top level yet which is a big problem for anyone using iframes. maybe this helps: https://stackoverflow.com/questions/1037839/how-to-force-link-from-iframe-to-be-opened-in-the-parent-window

alternatively we need to wait for the proper `<script>` embed which is more similar to calendly and other embeds. thats still in baby shoes but hopefully done soon!

**Comment by lharries** [comment]
Got it, thank you @PeerRich. Are there any solutions you'd recommend for detecting the Cal.com being booked and then moving the user onto the next page? Looking like we'll need to use websockets connected to the backend and monitoring the webhooks - which seems excessive

**Comment by hariombalhara** [comment]
> > Excited to see this PR!
> > 
> > We are currently working on a feature where we want to redirect the page based on a Cal.com booking being completed within an iFrame.
> > 
> > Does this PR enable an iFrame to trigger the redirection of the top level page? If not, I'd love to request it as a feature (ideally as part of this PR if not out of scope).
> > 
> > For implementation, one way could be to use `window.postMessage()` and allow the developer the handle the rest. This is what the embedded Calendly iFrame uses and gives the developer lots of flexibility including using dynamic success URLs and triggering appropriate analytics. Here's a link to how [Calendly does it](https://help.calendly.com/hc/en-us/articles/223147027-Embed-options-overview?tab=advanced#6)
> > 
> > Image of relevant Calendly docs
> 
> I havent tested but afaik it won't redirect at the top level yet which is a big problem for anyone using iframes. maybe this helps: https://stackoverflow.com/questions/1037839/how-to-force-link-from-iframe-to-be-opened-in-the-parent-window
> 
> alternatively we need to wait for the proper `<script>` embed which is more similar to calendly and other embeds. thats still in baby shoes but hopefully done soon!

It would actually redirect the parent of the cal iframe to the external URL.

**Comment by hariombalhara** [comment]
@PeerRich Maybe we can start sending postMessages to cal iframe parent. We can introduce `bookingSuccess` as the first event ?

**Comment by lharries** [comment]
> @PeerRich Maybe we can start sending postMessages to cal iframe parent. We can introduce `bookingSuccess` as the first event ?

This would work great!

**Comment by zomars** [comment]
I'll be updating this branch

File: packages/prisma/schema.prisma [hunk 3]
```diff
@@ -67,6 +67,7 @@ model EventType {
   currency                String                 @default("usd")
   slotInterval            Int?
   metadata                Json?
+  successRedirect         String?
```
Comment (zomars) [inline packages/prisma/schema.prisma:70]:
```suggestion
  successRedirectUrl         String?
```

**Review by PeerRich — APPROVED** [review]
awesome work! only waiting for merge conflicts to be resolved then merging

**Comment by hariombalhara** [comment]
Disabling auto merge. I found that in some earlier merge conflict resolution, the usage of SuccessRedirectEdit got removed. I have added it back but need to test once.

File: apps/web/server/routers/viewer/eventTypes.tsx [hunk 4]
```diff
@@ -18,6 +18,22 @@ function isPeriodType(keyInput: string): keyInput is PeriodType {
   return Object.keys(PeriodType).includes(keyInput);
 }
 
+/**
+ * Ensures that it is a valid HTTP URL
+ * It automatically avoids
+ * -  XSS attempts through javascript:alert('hi')
+ * - mailto: links
+ */
+function assertValidUrl(url) {
```
Comment (zomars) [inline apps/web/server/routers/viewer/eventTypes.tsx:27]:
```suggestion
function assertValidUrl(url: string) {
```

We're enabling strict mode soon so should avoid untyped parameters.

**Review by zomars — APPROVED** [review]

**Comment by hariombalhara** [comment]
This is ready to merge. Just migrations need to be run first @PeerRich @zomars

**Review by pumfleet — APPROVED** [review]

**Pull request merged**